drivers/adc/adc_mcp356xr.c

```diff
@@ -0,0 +1,1317 @@
+	if (err) {
+		return err;
+	}
+
+	err = adc_mcp356xr_set_inputs(&registers, channel_cfg->input_positive,
+				      channel_cfg->input_negative);
+	if (err) {
+		return err;
+	}
+
+	err = adc_mcp356xr_set_reference(&registers, channel_cfg->reference);
+	if (err) {
+		return err;
+	}
+
+	adc_mcp356xr_channel_registers_set(dev, channel_cfg->channel_id, registers);
+
+	return 0;
+}
+
+static int adc_mcp356xr_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	struct adc_mcp356xr_data *data = dev->data;
+
+	int err = 0;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	err = ads114s0x_adc_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_mcp356xr_read_async(const struct device *dev, const struct adc_sequence *sequence,
+				   struct k_poll_signal *async)
+{
+	struct adc_mcp356xr_data *data = dev->data;
+
+	int err = 0;
+
+	adc_context_lock(&data->ctx, true, async);
+	err = ads114s0x_adc_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+#endif
+
+static void adc_mcp356xr_acquisition_thread(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	struct adc_mcp356xr_data *data = p1;
+	const struct device *dev = data->dev;
+
+	while (1) {
+		k_sem_take(&data->start_sequence, K_FOREVER);
+
+		int err = 0;
+
+#ifndef CONFIG_ADC_MCP356XR_POLL
+		k_sem_reset(&data->irq_occurred);
+#endif
+
+		/* clear any leftover flags and sanitize state */
+		err = adc_mcp356xr_get_status_and_sanitize(dev, NULL, NULL, NULL);
+		if (err) {
+			LOG_ERR("Failed to sanitize ADC state (error = %d)", err);
+			adc_context_complete(&data->ctx, err);
+			continue;
+		}
+
+		err = adc_mcp356xr_run_acquisition_sequence(dev);
+		if (err) {
+			adc_context_complete(&data->ctx, err);
+			continue;
+		}
+
+		adc_context_on_sampling_done(&data->ctx, data->dev);
+	}
+}
+
+static int adc_mcp356xr_init(const struct device *dev)
+{
+	const struct adc_mcp356xr_config *config = dev->config;
+	struct adc_mcp356xr_data *data = dev->data;
+
+	if (!spi_is_ready_dt(&config->spi)) {
+		LOG_ERR("SPI device not ready");
+		return -ENODEV;
+	}
+
+	int err = 0;
+
+	err = adc_mcp356xr_reset_after_power_on(dev);
+	if (err) {
+		return err;
+	}
+
+#ifndef CONFIG_ADC_MCP356XR_POLL
+	err = adc_mcp356xr_irq_init(dev);
+	if (err) {
+		return err;
+	}
+#endif
+
+	k_sem_init(&data->start_sequence, 0, 1);
+
+	err = k_mutex_init(&data->channel_registers_mutex);
+	if (err) {
+		LOG_ERR("Failed to initialize internal mutex");
+	}
+
+	adc_context_init(&data->ctx);
+
+	k_tid_t tid =
+		k_thread_create(&data->thread, data->stack, K_KERNEL_STACK_SIZEOF(data->stack),
+				adc_mcp356xr_acquisition_thread, (void *)data, NULL, NULL,
+				CONFIG_ADC_MCP356XR_THREAD_PRIORITY, 0, K_NO_WAIT);
+
+	k_thread_name_set(tid, dev->name);
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+static const struct adc_driver_api adc_mcp356xr_api = {
```

Comment:
Device driver APIs need to be wrapped

```suggestion
static DEVICE_API(adc, adc_mcp356xr_api) = {
```

Reply:
Done